### PR TITLE
Make scheduler and duration optional for apis

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve? 
#171 
Make `scheduler` and `duration optional` for `chaos` APIs, then we can support this feature for each kind of `chaos` in parallel. 

### What is changed and how does it work?
Make `scheduler` and `duration` optional for `chaos` APIs, then we can support this feature for each kind of `chaos` in parallel.  `chaos-manager` would reject the reconciliation if the `scheduler` and `duration` are not both defined currently.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has Go code change

Side effects

 - None

### Does this PR introduce a user-facing change?:

```release-note
NONE
```

### CODE REVIEW

File: api/v1alpha1/podchaos_types.go [hunk 1]
```diff
@@ -53,13 +53,17 @@ func (in *PodChaos) IsDeleted() bool {
 	return !in.DeletionTimestamp.IsZero()
 }
 
-func (in *PodChaos) GetDuration() (time.Duration, error) {
-	duration, err := time.ParseDuration(in.Spec.Duration)
+//GetDuration would return the duration for chaos
+func (in *PodChaos) GetDuration() (*time.Duration, error) {
```
Comment (zhouqiang-cl) [inline api/v1alpha1/podchaos_types.go]: @cwen0 maybe we should refine this function, there are many repeat lines

File: config/crd/bases/pingcap.com_iochaos.yaml [hunk 2]
```diff
@@ -149,6 +149,13 @@ spec:
                   items:
                     type: string
                   type: array
+                phaseSelector:
+                  description: 'PodPhaseSelector is a set of condition of a pod at
```
Comment (zhouqiang-cl) [inline config/crd/bases/pingcap.com_iochaos.yaml]: s/PodPhaseSelector/IOPhaseSelector?

File: config/crd/bases/pingcap.com_networkchaos.yaml [hunk 3]
```diff
@@ -156,6 +156,13 @@ spec:
                   items:
                     type: string
                   type: array
+                phaseSelector:
+                  description: 'PodPhaseSelector is a set of condition of a pod at
```
Comment (zhouqiang-cl) [inline config/crd/bases/pingcap.com_networkchaos.yaml]: ditto

File: config/crd/bases/pingcap.com_networkchaos.yaml [hunk 4]
```diff
@@ -212,6 +219,13 @@ spec:
                       items:
                         type: string
                       type: array
+                    phaseSelector:
+                      description: 'PodPhaseSelector is a set of condition of a pod
```
Comment (zhouqiang-cl) [inline config/crd/bases/pingcap.com_networkchaos.yaml]: ditto

**Comment by zhouqiang-cl** [comment]
LGTM

**Comment by zhouqiang-cl** [comment]
@cwen0 @YangKeao @mahjonp PTAL

File: api/v1alpha1/iochaos_types.go [hunk 5]
```diff
@@ -84,7 +84,7 @@ type IoChaosSpec struct {
 	// such as "300ms", "-1.5h" or "2h45m".
 	// Valid time units are "ns", "us" (or "µs"), "ms", "s", "m", "h".
 	// +optional
-	Duration string `json:"duration"`
+	Duration *string `json:"duration,omitempty"`
```
Comment (siddontang) [inline api/v1alpha1/iochaos_types.go]: why using pointer here? you can use empty to check the duration directly

File: api/v1alpha1/iochaos_types.go [hunk 6]
```diff
@@ -175,13 +175,17 @@ func (in *IoChaos) IsDeleted() bool {
 	return !in.DeletionTimestamp.IsZero()
 }
 
-func (in *IoChaos) GetDuration() (time.Duration, error) {
-	duration, err := time.ParseDuration(in.Spec.Duration)
+//GetDuration would return the duration for chaos
+func (in *IoChaos) GetDuration() (*time.Duration, error) {
+	if in.Spec.Duration == nil {
+		return nil, nil
```
Comment (siddontang) [inline api/v1alpha1/iochaos_types.go]: please format your comment

File: api/v1alpha1/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -175,7 +175,16 @@ func (in *IoChaosList) DeepCopyObject() runtime.Object {
 func (in *IoChaosSpec) DeepCopyInto(out *IoChaosSpec) {
```
Comment (siddontang) [inline api/v1alpha1/zz_generated.deepcopy.go]: seem you can use https://github.com/jinzhu/copier to simplify the codes

File: api/v1alpha1/iochaos_types.go [hunk 8]
```diff
@@ -175,13 +175,17 @@ func (in *IoChaos) IsDeleted() bool {
...
+	}
+	zeroHour := 0 * time.Hour
+	duration, err := time.ParseDuration(*in.Spec.Duration)
 	if err != nil {
-		return time.Hour * 0, err
+		return &zeroHour, err
```
Comment (mahjonp) [inline api/v1alpha1/iochaos_types.go]:
```suggestion
		return nil, err
```

File: api/v1alpha1/networkchaos_types.go [hunk 9]
```diff
@@ -189,13 +189,17 @@ func (in *NetworkChaos) IsDeleted() bool {
...
+//GetDuration would return the duration for chaos
+func (in *NetworkChaos) GetDuration() (*time.Duration, error) {
+	if in.Spec.Duration == nil {
+		return nil, nil
+	}
+	zeroHour := 0 * time.Hour
+	duration, err := time.ParseDuration(*in.Spec.Duration)
 	if err != nil {
-		return time.Hour * 0, err
+		return &zeroHour, err
```
Comment (mahjonp) [inline api/v1alpha1/networkchaos_types.go]: ditto

File: api/v1alpha1/common_types.go [hunk 10]
```diff
@@ -55,7 +55,7 @@ type SelectorSpec struct {
 	// +optional
 	AnnotationSelectors map[string]string `json:"annotationSelectors,omitempty"`
 
-	// PodPhaseSelector is a set of condition of a pod at the current time.
+	// PhaseSelector is a set of condition of a pod at the current time.
 	// supported value: Pending / Running / Succeeded / Failed / Unknown
 	PodPhaseSelectors []string `json:"phaseSelector,omitempty"`
```
Comment (mahjonp) [inline api/v1alpha1/common_types.go]:
```suggestion
	PodPhaseSelectors []string `json:"phaseSelectors,omitempty"`
```

File: config/crd/bases/pingcap.com_iochaos.yaml [hunk 11]
```diff
@@ -149,6 +149,13 @@ spec:
                   items:
                     type: string
                   type: array
+                phaseSelector:
```
Comment (mahjonp) [inline config/crd/bases/pingcap.com_iochaos.yaml]: it's better to  replace with `phaseSelectors`?

File: manifests/crd.yaml [hunk 12]
```diff
@@ -79,11 +79,11 @@ spec:
                 one / all / fixed / fixed-percent / random-max-percent'
               type: string
             nextRecover:
-              description: Next time when this action will be recovered
+              description: Next time when this action will be recovered.
```
Comment (mahjonp) [inline manifests/crd.yaml]: revert these changes?

File: config/crd/bases/pingcap.com_iochaos.yaml (hunk 11 above)
Comment (Yisaer) [reply config/crd/bases/pingcap.com_iochaos.yaml]: This is auto-generated by executing `make manifests`

File: manifests/crd.yaml (hunk 12 above)
Comment (Yisaer) [reply manifests/crd.yaml]: This is auto-generated by executing `make yaml`.

File: api/v1alpha1/iochaos_types.go (hunk 5 above)
Comment (Yisaer) [reply api/v1alpha1/iochaos_types.go]: Using pointer here is to make chaos API more user-friendly. If users need the chaos action always perform after they apply the chaos API, it is more reasonable that they omit the `scheduler` and the `duration` instead of leaving those as empty.

File: api/v1alpha1/zz_generated.deepcopy.go (hunk 7 above)
Comment (Yisaer) [reply api/v1alpha1/zz_generated.deepcopy.go]: This is auto-generated, so I suggest making it be as it generated by scripts.

**Comment by Yisaer** [comment]
In this request, I also execute `make yaml` after I finish editing the API spec and I found that the current CRD isn't the latest which I mentioned in https://github.com/pingcap/chaos-mesh/issues/173, so we can see some changes in CRD file which is not related in this request.

File: api/v1alpha1/iochaos_types.go (hunk 6 above)
Comment (Yisaer) [reply api/v1alpha1/iochaos_types.go]: updated.

File: api/v1alpha1/iochaos_types.go (hunk 8 above)
Comment (Yisaer) [reply api/v1alpha1/iochaos_types.go]: updated.

File: api/v1alpha1/networkchaos_types.go (hunk 9 above)
Comment (Yisaer) [reply api/v1alpha1/networkchaos_types.go]: updated.

File: api/v1alpha1/common_types.go (hunk 10 above)
Comment (Yisaer) [reply api/v1alpha1/common_types.go]: I think this change would break the backward compatible.

File: api/v1alpha1/common_types.go [hunk 13]
```diff
@@ -55,7 +55,7 @@ type SelectorSpec struct {
 	// +optional
 	AnnotationSelectors map[string]string `json:"annotationSelectors,omitempty"`
 
-	// PodPhaseSelector is a set of condition of a pod at the current time.
+	// PhaseSelector is a set of condition of a pod at the current time.
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply api/v1alpha1/common_types.go]: updaetd.

File: api/v1alpha1/iochaos_types.go [hunk 14]
```diff
@@ -175,13 +175,17 @@ func (in *IoChaos) IsDeleted() bool {
 	return !in.DeletionTimestamp.IsZero()
 }
 
-func (in *IoChaos) GetDuration() (time.Duration, error) {
-	duration, err := time.ParseDuration(in.Spec.Duration)
+//GetDuration would return the duration for chaos
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply api/v1alpha1/iochaos_types.go]: updated.

File: api/v1alpha1/common_types.go (hunk 10 above)
Comment (Yisaer) [reply api/v1alpha1/common_types.go]: I think this change would break the backward compatible.

Comment (mahjonp) [reply api/v1alpha1/common_types.go]: I guess this field hasn't been used now.

Comment (Yisaer) [reply api/v1alpha1/common_types.go]: It's okay to me if this hasn' been used.

File: api/v1alpha1/common_types.go [hunk 15]
```diff
@@ -55,9 +55,9 @@ type SelectorSpec struct {
 	// +optional
 	AnnotationSelectors map[string]string `json:"annotationSelectors,omitempty"`
 
-	// PodPhaseSelector is a set of condition of a pod at the current time.
+	// PodPhaseSelectors is a set of condition of a pod at the current time.
 	// supported value: Pending / Running / Succeeded / Failed / Unknown
-	PodPhaseSelectors []string `json:"phaseSelector,omitempty"`
+	PodPhaseSelectors []string `json:"phaseSelectors,omitempty"`
```
Comment (cwen0) [inline api/v1alpha1/common_types.go]: `json:"podPhaseSelectors, omitempty"`

Comment (Yisaer) [reply api/v1alpha1/common_types.go]: updated.
